engines/osxnative/sandbox.go

```diff
@@ -0,0 +1,147 @@
+package osxnative
+
+import (
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"io"
+	"mime"
+	"net/http"
+	"os"
+	"os/exec"
+)
+
+// white list of environment variables that must
+// be passed to the child process
+var environmentWhitelist = []string{
+	"PATH",
+	"HOME",
+	"USER",
+	"SHELL",
+	"TMPDIR",
+	"PWD",
+	"EDITOR",
+	"LANG",
+	"LOGNAME",
+	"TERM",
+	"TERM_PROGRAM",
+}
+
+type sandbox struct {
+	context     *runtime.TaskContext
+	taskPayload *payload
+	env         *[]string
+	aborted     bool
+}
+
+func newSandbox(context *runtime.TaskContext, taskPayload *payload, env *[]string) *sandbox {
+	return &sandbox{
+		context,
+		taskPayload,
+		env,
+		false,
+	}
+}
+
+func downloadLink(link string) (string, error) {
+	resp, err := http.Get(link)
+
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	contentDisposition := resp.Header.Get("Content-Disposition")
+	_, params, err := mime.ParseMediaType(contentDisposition)
+
+	if err != nil {
+		return "", err
+	}
+
+	filename := params["filename"]
+	file, err := os.Create(filename)
+
+	if err != nil {
+		return "", err
+	}
+
+	defer file.Close()
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		os.Remove(filename)
+		return "", err
+	}
+
+	return filename, nil
+}
+
+func (s *sandbox) WaitForResult() (engines.ResultSet, error) {
+	if s.aborted {
+		return resultset{success: false}, engines.ErrSandboxAborted
+	}
+
+	if s.taskPayload.Link != "" {
+		filename, err := downloadLink(s.taskPayload.Link)
+
+		if err != nil {
+			s.context.LogError(err)
+			return nil, engines.ErrNonFatalInternalError
+		}
+
+		defer os.Remove(filename)
+
+		err = os.Chmod(filename, 0777)
+
+		if err != nil {
+			s.context.LogError(err)
+			return nil, engines.ErrNonFatalInternalError
+		}
+	}
+
+	env := make([]string, len(*s.env), len(*s.env)+len(environmentWhitelist))
+	copy(env, *s.env)
+
+	// Use the host environment plus custom environment variables
+	for _, e := range environmentWhitelist {
+		value := os.Getenv(e)
+		if value != "" {
+			env = append(env, e+"="+value)
+		}
+	}
+
+	cmd := exec.Command(s.taskPayload.Command[0], s.taskPayload.Command[1:]...)
+	cmd.Env = env
+
+	output, err := cmd.CombinedOutput()
+	s.context.Log(string(output))
+
+	if err != nil {
+		s.context.LogError("Command \"", s.taskPayload.Command, "\" failed to run: ", err)
+		switch err.(type) {
+		case *exec.ExitError:
+			return resultset{success: false}, nil
+		default:
+			return resultset{success: false}, engines.ErrNonFatalInternalError
+		}
+	}
+
+	return resultset{success: true}, nil
+}
+
+func (s *sandbox) Abort() error {
+	s.aborted = true
+	return nil
+}
+
+func (*sandbox) NewShell() (engines.Shell, error) {
```

Comment:
It took a while for me to understand, but if I use `SandboxBase`, some methods will be implemented with an object receiver, others with a pointer receiver, so neither will completely implement the `Sandbox`interface. Kinda weird.